### PR TITLE
test(secretserver): improve test coverage for SecretServer provider

### LINKED ISSUE #5640: Improve test coverage for SecretServer provider
## Problem
The SecretServer provider currently has inadequate test coverage at 76.3% overall, with several critical functions having 0% coverage.

### Functions with 0% Coverage (Critical)
- `PushSecret` - 0.0%
- `DeleteSecret` - 0.0%
- `SecretExists` - 0.0%
- `Validate` - 0.0%
- `GetSecretMap` - 0.0%
- `GetAllSecrets` - 0.0%
- `Close` - 0.0%
- `Capabilities` - 0.0%
- `NewProvider` - 0.0%
- `ProviderSpec` - 0.0%
- `MaintenanceStatus` - 0.0%

### Functions with Partial Coverage
- `GetSecret` - 95.7% (good, but could reach 100%)
- `NewClient` - 91.7%
- `getSecret` (private) - 87.5%
- `getConfig` - 94.7%
- `loadConfigSecret` - 80.0%
- `validateStoreSecretRef` - 75.0%

### Functions with Full Coverage ✓
- `validateSecretRef` - 100%
- `doesConfigDependOnNamespace` - 100%
- `ValidateStore` - 100%

## Impact
The lack of test coverage for core operations like `PushSecret`, `DeleteSecret`, and `SecretExists` poses risks for:
- Regression bugs going undetected
- Difficulty refactoring code safely
- Reduced confidence in provider reliability
- Harder maintenance and debugging

## Proposed Solution
Add comprehensive unit tests for all uncovered functions, prioritizing:

1. **High Priority**: CRUD operations (`PushSecret`, `DeleteSecret`, `SecretExists`)
2. **Medium Priority**: Validation and utility functions (`Validate`, `GetSecretMap`, `GetAllSecrets`)
3. **Low Priority**: Lifecycle and metadata functions (`Close`, `Capabilities`, `NewProvider`, `ProviderSpec`, `MaintenanceStatus`)

## Coverage Data
```
total: (statements) 76.3%
```

<details>
<summary>Full coverage report</summary>
```
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:50:    GetSecret        95.7%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:96:    PushSecret        0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:101:   DeleteSecret      0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:106:   SecretExists      0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:111:   Validate          0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:115:   GetSecretMap      0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:138:   GetAllSecrets     0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:142:   Close             0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/client.go:147:   getSecret        87.5%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:54:  Capabilities      0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:59:  NewClient        91.7%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:119: loadConfigSecret 80.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:134: validateStoreSecretRef 75.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:143: validateSecretRef 100.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:160: doesConfigDependOnNamespace 100.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:170: getConfig        94.7%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:203: ValidateStore    100.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:209: NewProvider       0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:214: ProviderSpec      0.0%
github.com/external-secrets/external-secrets/providers/v1/secretserver/provider.go:221: MaintenanceStatus 0.0%
total:                                                                                  (statements)     76.3%
```

</details>

## Additional Context
- Provider location: `providers/v1/secretserver/`
- Files affected: `client.go`, `provider.go`

### PR DESCRIPTION
**Pull request by DelineaSahilWankhede**
## Problem Statement
The SecretServer provider had insufficient test coverage at 76.3% overall, with 11 critical functions having 0% coverage. This included core CRUD operations like `PushSecret`, `DeleteSecret`, and `SecretExists`, as well as validation and lifecycle functions. The lack of tests posed risks for regression bugs, made refactoring difficult, and reduced confidence in the provider's reliability.

## Related Issue
Fixes #5640 

## Proposed Changes
This PR adds comprehensive unit tests for previously uncovered functions in the SecretServer provider:

### Added Tests
- **CRUD Operations**: `PushSecret`, `DeleteSecret`, `SecretExists` - tests for success cases, error handling, and edge cases
- **Data Retrieval**: `GetSecretMap`, `GetAllSecrets` - tests for multiple secret handling and data mapping
- **Validation**: `Validate` - tests for configuration validation logic
- **Lifecycle**: `Close` - tests for proper resource cleanup
- **Provider Metadata**: `Capabilities`, `NewProvider`, `ProviderSpec`, `MaintenanceStatus` - tests for provider initialization and metadata

### Approach
- Used table-driven tests for better coverage of multiple scenarios
- Added mock implementations where needed to isolate units under test
- Focused on both happy paths and error conditions
- Improved coverage of existing partially tested functions

### Coverage Improvement
- **Before**: 76.3% overall coverage (11 functions at 0%)
- **After**: 93.2% overall coverage (all previously uncovered functions now at 100%)

### Detailed Coverage Changes
| Function | Before | After | Status |
|----------|--------|-------|--------|
| `PushSecret` | 0.0% | 100.0% | ✅ |
| `DeleteSecret` | 0.0% | 100.0% | ✅ |
| `SecretExists` | 0.0% | 100.0% | ✅ |
| `Validate` | 0.0% | 100.0% | ✅ |
| `GetAllSecrets` | 0.0% | 100.0% | ✅ |
| `Close` | 0.0% | 100.0% | ✅ |
| `Capabilities` | 0.0% | 100.0% | ✅ |
| `NewProvider` | 0.0% | 100.0% | ✅ |
| `ProviderSpec` | 0.0% | 100.0% | ✅ |
| `GetSecretMap` | 0.0% | 93.3% | ✅ |
| `getSecret` | 87.5% | 93.8% | ✅ Improved |
| `GetSecret` | 95.7% | 95.7% | ✓ Maintained |

## Format
Please ensure that your PR follows the following format for the title:
```
feat(scope): add new feature
fix(scope): fix bug
docs(scope): update documentation
chore(scope): update build tool or dependencies
ref(scope): refactor code
clean(scope): provider cleanup
test(scope): add tests
perf(scope): improve performance
desig(scope): improve design
```
Where `scope` is _optionally_ one of:
- charts
- release
- testing
- security
- templating

## Checklist
- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: providers/v1/secretserver/client.go [hunk 1]
```diff
@@ -117,6 +117,11 @@ func (c *client) GetSecretMap(ctx context.Context, ref esv1.ExternalSecretDataRe
 	if err != nil {
 		return nil, err
 	}
+	// Ensure secret has fields before indexing into them
+	if secret.Fields == nil || len(secret.Fields) == 0 {
+		return nil, errors.New("secret contains no fields")
+	}
+
```
Comment (gusfcarvalho) [inline providers/v1/secretserver/client.go:126]: shouldn't this return a `NoSecretError` instead? It is an error type we have in external-secrets to make provider calls compatible with `CreationPolicy: Merge` and `DeletionPolicy: Merge/Delete`

Comment (gusfcarvalho) [reply providers/v1/secretserver/client.go:126]: would possibly apply to other places, which I reckon might be out of scope of this PR - so a nit really

**Review by gusfcarvalho — APPROVED** [review]

File: providers/v1/secretserver/client.go (hunk 1 above)
Comment (DelineaSahilWankhede) [reply providers/v1/secretserver/client.go:126]: Thanks for the suggestion!
Since this PR is focused on test coverage improvements, I'll keep the current behavior.
I can open a separate issue/PR to replace plain errors with `NoSecretError` across the provider.

File: providers/v1/secretserver/client_test.go [hunk 2]
```diff
@@ -324,3 +324,271 @@ func TestGetSecretSecretServer(t *testing.T) {
 		})
 	}
 }
+
+// TestGetSecret_JSONMarshalFailure tests GetSecret when json.Marshal fails
+func TestGetSecret_JSONMarshalFailure(t *testing.T) {
+	ctx := context.Background()
```
Comment (Skarlso) [inline providers/v1/secretserver/client_test.go]: Please use t.Context().

File: providers/v1/secretserver/client_test.go [hunk 3]
```diff
@@ -324,3 +324,271 @@ func TestGetSecretSecretServer(t *testing.T) {
...
+		{
+			FieldName: "Foo",
+			ItemValue: string([]byte{0xff, 0xfe}), // invalid UTF-8 → forces marshal failure
+		},
+	}
+
+	// GetSecret calls getSecret which returns the secret, so no error expected
+	_, err := c.GetSecret(ctx, esv1.ExternalSecretDataRemoteRef{Key: "0"})
+	// The secret is found but ItemValue is invalid, so it should not error at getSecret level
+	assert.NoError(t, err)
```
Comment (Skarlso) [inline providers/v1/secretserver/client_test.go]: Please use `requires`.

File: providers/v1/secretserver/client_test.go [hunk 4]
```diff
@@ -324,3 +324,271 @@ func TestGetSecretSecretServer(t *testing.T) {
 		})
 	}
 }
+
+// TestGetSecret_JSONMarshalFailure tests GetSecret when json.Marshal fails
+func TestGetSecret_JSONMarshalFailure(t *testing.T) {
```
Comment (Skarlso) [inline providers/v1/secretserver/client_test.go]: Don't use `_` in test function names.

File: providers/v1/secretserver/client_test.go [hunk 5]
```diff
@@ -324,3 +324,271 @@ func TestGetSecretSecretServer(t *testing.T) {
...
+
+	// GetSecret calls getSecret which returns the secret, so no error expected
+	_, err := c.GetSecret(ctx, esv1.ExternalSecretDataRemoteRef{Key: "0"})
+	// The secret is found but ItemValue is invalid, so it should not error at getSecret level
+	assert.NoError(t, err)
+}
+
+// TestGetSecret_EmptySecretsList tests GetSecret when the secrets list is empty
+func TestGetSecret_EmptySecretsList(t *testing.T) {
+	ctx := context.Background()
```
Comment (Skarlso) [inline providers/v1/secretserver/client_test.go:379]: Here as well and every other test.

File: providers/v1/secretserver/provider_test.go [hunk 6]
```diff
@@ -546,3 +546,180 @@ func generateRandomString() string {
...
+			assert.NotNil(t, got)
+			assert.NotNil(t, got.SecretServer)
+			assert.IsType(t, tc.wantType, got)
+
+			// Edge: call ProviderSpec multiple times
+			got2 := ProviderSpec()
+			assert.IsType(t, tc.wantType, got2)
+		})
+	}
+}
```
Comment (Skarlso) [inline providers/v1/secretserver/provider_test.go]: This test has no value aside from testing _lines_ which I don't believe add anything.

File: providers/v1/secretserver/provider_test.go [hunk 7]
```diff
@@ -546,3 +546,180 @@ func generateRandomString() string {
...
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := MaintenanceStatus()
+			assert.Equal(t, tc.want, got)
+
+			// Edge: call MaintenanceStatus on nil
+			assert.Equal(t, esv1.MaintenanceStatusMaintained, MaintenanceStatus())
+		})
+	}
+}
```
Comment (Skarlso) [inline providers/v1/secretserver/provider_test.go]: Also is just there to increase coverage % which is a meaningless indicator of "quality".

File: providers/v1/secretserver/provider_test.go (hunk 6 above)
Comment (DelineaSahilWankhede) [reply providers/v1/secretserver/provider_test.go]: > This test has no value aside from testing _lines_ which I don't believe add anything.

I updated `TestProviderSpec` to assert `ProviderSpec()` returns independent instances (I mutate the returned object and verify a subsequent call is unaffected). This prevents regressions where `ProviderSpec` could accidentally return a shared/mutable object rather than a fresh spec.

File: providers/v1/secretserver/provider_test.go (hunk 7 above)
Comment (DelineaSahilWankhede) [reply providers/v1/secretserver/provider_test.go]: > Also is just there to increase coverage % which is a meaningless indicator of "quality".

I removed `TestMaintenanceStatus` because it only asserted a constant return value and didn’t exercise any behavior; as suggested I avoided keeping tests that exist solely to inflate coverage.

**Review by DelineaSahilWankhede — COMMENTED** [review]
Hi @Skarlso @gusfcarvalho 
The SonarQube check is flagging low-priority issues about removing the "Get" prefix from function names (`GetSecret`, `GetSecretMap`, `GetAllSecrets`). 
However, these are interface method implementations from `esv1.SecretsClient` and cannot be renamed without breaking the interface contract. 
These warnings are making the Quality Gate fail.
How should we handle this?

**Comment by Skarlso** [comment]
We'll have to ignore those. I'll fix those. They sprung up again, because you touched the package.

**Comment by Skarlso** [comment]
Please take care of the lint and check-diff issues then it's okay.

**Comment by DelineaSahilWankhede** [comment]
> We'll have to ignore those. I'll fix those. They sprung up again, because you touched the package.

Thank you. As per suggestion i have taken care of the lint and check-diff issues.
